src/python/pants/backend/jvm/subsystems/scala_platform.py

```diff
@@ -132,17 +138,37 @@ def register_custom_tool(key):
     register_custom_tool('scala-repl')
     register_custom_tool('scalastyle')
 
-  def _tool_classpath(self, tool, products):
+  def _tool_classpath(self, tool, products, scheduler):
     """Return the proper classpath based on products and scala version."""
-    return self.tool_classpath_from_products(products,
-                                             self._key_for_tool_version(tool, self.version),
-                                             scope=self.options_scope)
+    classpath = self.tool_classpath_from_products(products,
+                                                  self._key_for_tool_version(tool, self.version),
+                                                  scope=self.options_scope)
+    classpath = tuple(fast_relpath(c, get_buildroot()) for c in classpath)
+
+    return self._memoized_scalac_classpath(classpath, scheduler)
+
+  @memoized_method
+  def _memoized_scalac_classpath(self, scala_path, scheduler):
+    if scala_path:
+      snapshots = scheduler.capture_snapshots(tuple(PathGlobsAndRoot(PathGlobs([path]), get_buildroot()) for path in scala_path))
+      return [ClasspathEntry(path, snapshot) for path, snapshot in list(zip(scala_path, snapshots))]
+    else:
+      return [ClasspathEntry(p, None) for p in scala_path]
 
+  @deprecated('1.12.0.dev0', 'Use compiler_classpath_entries instead.')
   def compiler_classpath(self, products):
-    return self._tool_classpath('scalac', products)
+    """Returns classpath as paths."""
+    compiler_classpath_entries = self._tool_classpath('scalac', products)
+    return [classpath_entry.path for classpath_entry in compiler_classpath_entries]
+
+  def compiler_classpath_entries(self, products, scheduler):
+    """Returns classpath entries for the scalac tool."""
+    return self._tool_classpath('scalac', products, scheduler)
 
   def style_classpath(self, products):
-    return self._tool_classpath('scalastyle', products)
+    """Returns classpath as paths for scalastyle."""
+    classpath_entries = self._tool_classpath('scalastyle', products)
+    return [classpath_entry.path for classpath_entry in classpath_entries]
```

Comment:
Since these aren't API public, I think you're good to inline them, if you want. Maintaining them this way though may be easier. Up to you!

Reply:
I prefer it this way.